core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -0,0 +1,243 @@
+package filetransfer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/googleapis/gax-go/v2"
+	"github.com/wandb/wandb/core/pkg/observability"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/iterator"
+)
+
+type GCSClient interface {
+	Bucket(name string) *storage.BucketHandle
+	SetRetry(opts ...storage.RetryOption)
+}
+
+// GCSFileTransfer uploads or downloads files to/from GCS
+type GCSFileTransfer struct {
+	// client is the HTTP client for the file transfer
+	client GCSClient
+
+	// logger is the logger for the file transfer
+	logger *observability.CoreLogger
+
+	// fileTransferStats is used to track upload/download progress
+	fileTransferStats FileTransferStats
+
+	// background context is used to create a reader and get the client
+	ctx context.Context
+}
+
+const maxWorkers int = 1000
+
+// NewGCSFileTransfer creates a new fileTransfer
+func NewGCSFileTransfer(
+	client GCSClient,
+	logger *observability.CoreLogger,
+	fileTransferStats FileTransferStats,
+) *GCSFileTransfer {
+	ctx := context.Background()
+	if client == nil {
+		var err error
+		client, err = storage.NewClient(ctx)
+		if err != nil {
+			logger.CaptureError(fmt.Errorf("gcs file transfer: error creating new gcs client: %v", err))
+			return nil
```

Comment:
We shouldn't be returning nil here, rather the actual error thrown by storage.NewClient or something more specific.